Return 100-continue response before accepting the body if expected

g4s8 opened the pull request:
artipie/artipie#382 - send `100 Continue` response status before accepting request body if `Expect: 100-continue` header found.

g4s8 committed:
artipie/artipie#382 - 100 continue if expected

olenagerasimova commented on src/main/java/com/artipie/vertx/VertxSliceServer.java:
```diff
@@ -165,6 +165,9 @@ public void close() {
      * @return Completion of request serving.
      */
     private CompletionStage<Void> serve(final HttpServerRequest req) {
+        if (expectContinue(req)) {
```
@g4s8 what about test?

olenagerasimova reviewed (changes requested):
@g4s8 please check question below

Sammers21 commented on src/main/java/com/artipie/vertx/VertxSliceServer.java:
```diff
@@ -165,6 +165,9 @@ public void close() {
      */
     private CompletionStage<Void> serve(final HttpServerRequest req) {
+        if (expectContinue(req)) {
+            req.response().writeContinue();
```
@g4s8, you need to return from the function since the response is already produced. And a unit test would be welcome. Not sure if that would work as it is now.

Sammers21 reviewed (commented):
@g4s8, I am also in doubt if it's a good place to land these changes. The server is just a backend for the slice interface and I'm not sure if it should contain any business logic code. Maybe it's better to do it artipie/artipie code? If you would change the underlying server one day, we would lose the code you wrote, i.e. you would need to write it again.

g4s8 committed:
artipie/artipie#382 - refactored slice server connection

g4s8 commented:
@olenagerasimova @Sammers21 thanks, I've fixed your comments and refactored `Connection` implementation to process `100` status from `Slice`s response correctly. I've checked it locally as snapshot version for `artipie/artipie` and it's working fine. The correct HTTP flow of `100` status is:
```
> PUT /artipie/test/com/artipie/sample-for-deployment/maven-metadata.xml HTTP_1_1
> Cache-control: no-cache
> Cache-store: no-store
> Pragma: no-cache
> User-Agent: Apache-Maven/3.6.2 (Java 13.0.1; Linux 5.4.38-gentoo-x86_64)
> Content-Length: 333
> Host: localhost:8080
> Connection: Keep-Alive
> Expect: 100-continue
> Accept-Encoding: gzip,deflate 
>
< 100 Continue
> (request body)
< 201 Created
```
We're sending `100` status after request head but before request body, then we're processing the body, and sending final status code (`201` in example) and response if needed.

Sammers21 commented:
@g4s8, so can you please close this PR?

g4s8 commented:
> @g4s8, so can you please close this PR?

@Sammers21 why to close? We need `100` status support in server implementation, Vertx won't allow to send two statuses one by one, it has special method for sending `100` status `response.writeContinue()`, so we can't avoid PRs in `artipie/vertx-server`.

g4s8 commented:
@rultor release, tag=`0.3.2`

rultor commented:
> @rultor release, tag=`0.3.2`

@g4s8 OK, I will release it now. Please check the progress [here](https://www.rultor.com/t/22766-665713525)

rultor commented:
> @rultor release, tag=`0.3.2`

@g4s8 Oops, I failed. You can see the full log [here](https://www.rultor.com/t/22766-665713525) (spent 16s)

```
+ set -e
+ set -o pipefail
++ dirname ./run.sh
+ cd .
+ echo 16379
+ echo '1.69.1 839260fab'
1.69.1 839260fab
+ date
Wed Jul 29 16:55:38 CEST 2020
+ uptime
 16:55:38 up 286 days, 13:04,  0 users,  load average: 5.10, 4.65, 4.62
+ cat
.rultor.yml in yegor256/home doesn't allow artipie/vertx-server to use its assets (there are 107 friends), see http://doc.rultor.com/reference.html#assets
+ exit -1
'cid' file is absent, container wasn't started correctly
```

g4s8 commented:
@rultor release, tag=`0.3.2`

rultor commented:
> @rultor release, tag=`0.3.2`

@g4s8 OK, I will release it now. Please check the progress [here](https://www.rultor.com/t/22766-665715668)

rultor commented:
> @rultor release, tag=`0.3.2`

@g4s8 Oops, I failed. You can see the full log [here](https://www.rultor.com/t/22766-665715668) (spent 16s)

```
+ set -e
+ set -o pipefail
++ dirname ./run.sh
+ cd .
+ echo 16944
+ echo '1.69.1 839260fab'
1.69.1 839260fab
+ date
Wed Jul 29 16:59:43 CEST 2020
+ uptime
 16:59:43 up 286 days, 13:08,  0 users,  load average: 4.77, 4.74, 4.67
+ cat
Since @Kirill is the modifier of .rultor.yml in artipie/vertx-server, that's why assets are not permitted to use in yegor256/artipie-keys
+ exit -1
'cid' file is absent, container wasn't started correctly
```

0crat commented:
Job `gh:artipie/vertx-server#18` is not assigned, can't get performer

<!-- https://www.0crat.com/footprint/CT2E6TK9B/4150c458-4dfc-4681-83f8-8fd0cf8b3176, version: 0.54.5, hash: ${buildNumber} -->

g4s8 commented:
@rultor release, tag=`0.3.2`

rultor commented:
> @rultor release, tag=`0.3.2`

@g4s8 OK, I will release it now. Please check the progress [here](https://www.rultor.com/t/22766-666177468)

rultor commented:
> @rultor release, tag=`0.3.2`

@g4s8 Oops, I failed. You can see the full log [here](https://www.rultor.com/t/22766-666177468) (spent 12min)

```
    \u001b[1mat\u001b[m org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (\u001b[1mLifecycleModuleBuilder.java:117\u001b[m)
    \u001b[1mat\u001b[m org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (\u001b[1mLifecycleModuleBuilder.java:81\u001b[m)
    \u001b[1mat\u001b[m org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (\u001b[1mSingleThreadedBuilder.java:56\u001b[m)
    \u001b[1mat\u001b[m org.apache.maven.lifecycle.internal.LifecycleStarter.execute (\u001b[1mLifecycleStarter.java:128\u001b[m)
    \u001b[1mat\u001b[m org.apache.maven.DefaultMaven.doExecute (\u001b[1mDefaultMaven.java:305\u001b[m)
    \u001b[1mat\u001b[m org.apache.maven.DefaultMaven.doExecute (\u001b[1mDefaultMaven.java:192\u001b[m)
    \u001b[1mat\u001b[m org.apache.maven.DefaultMaven.execute (\u001b[1mDefaultMaven.java:105\u001b[m)
    \u001b[1mat\u001b[m org.apache.maven.cli.MavenCli.execute (\u001b[1mMavenCli.java:956\u001b[m)
    \u001b[1mat\u001b[m org.apache.maven.cli.MavenCli.doMain (\u001b[1mMavenCli.java:288\u001b[m)
    \u001b[1mat\u001b[m org.apache.maven.cli.MavenCli.main (\u001b[1mMavenCli.java:192\u001b[m)
    \u001b[1mat\u001b[m jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (\u001b[1mNative Method\u001b[m)
    \u001b[1mat\u001b[m jdk.internal.reflect.NativeMethodAccessorImpl.invoke (\u001b[1mNativeMethodAccessorImpl.java:62\u001b[m)
    \u001b[1mat\u001b[m jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (\u001b[1mDelegatingMethodAccessorImpl.java:43\u001b[m)
    \u001b[1mat\u001b[m java.lang.reflect.Method.invoke (\u001b[1mMethod.java:566\u001b[m)
    \u001b[1mat\u001b[m org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (\u001b[1mLauncher.java:282\u001b[m)
    \u001b[1mat\u001b[m org.codehaus.plexus.classworlds.launcher.Launcher.launch (\u001b[1mLauncher.java:225\u001b[m)
    \u001b[1mat\u001b[m org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (\u001b[1mLauncher.java:406\u001b[m)
    \u001b[1mat\u001b[m org.codehaus.plexus.classworlds.launcher.Launcher.main (\u001b[1mLauncher.java:347\u001b[m)
\u001b[1mCaused by\u001b[m: org.sonatype.nexus.client.core.exception.NexusClientResponseException: \u001b[1;31mGateway Timeout : entity body dump follows: <html>

<head><title>504 Gateway Time-out</title></head>

<body bgcolor="white">

<center><h1>504 Gateway Time-out</h1></center>

</body>

</html>

\u001b[m
    \u001b[1mat\u001b[m com.sonatype.nexus.staging.client.internal.ExceptionConverter.handleUniformInterfaceException (\u001b[1mExceptionConverter.java:82\u001b[m)
    \u001b[1mat\u001b[m com.sonatype.nexus.staging.client.internal.ExceptionConverter.runAndReturn (\u001b[1mExceptionConverter.java:41\u001b[m)
    \u001b[1mat\u001b[m com.sonatype.nexus.staging.client.internal.StagingWorkflowV2ServiceImpl.startStaging (\u001b[1mStagingWorkflowV2ServiceImpl.java:175\u001b[m)
    \u001b[1mat\u001b[m org.sonatype.nexus.maven.staging.deploy.strategy.AbstractStagingDeployStrategy.beforeUpload (\u001b[1mAbstractStagingDeployStrategy.java:103\u001b[m)
    \u001b[1mat\u001b[m org.sonatype.nexus.maven.staging.deploy.strategy.StagingDeployStrategy.finalizeDeploy (\u001b[1mStagingDeployStrategy.java:122\u001b[m)
    \u001b[1mat\u001b[m org.sonatype.nexus.maven.staging.deploy.DeployMojo.execute (\u001b[1mDeployMojo.java:213\u001b[m)
    \u001b[1mat\u001b[m org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (\u001b[1mDefaultBuildPluginManager.java:137\u001b[m)
    \u001b[1mat\u001b[m org.apache.maven.lifecycle.internal.MojoExecutor.execute (\u001b[1mMojoExecutor.java:210\u001b[m)
    \u001b[1mat\u001b[m org.apache.maven.lifecycle.internal.MojoExecutor.execute (\u001b[1mMojoExecutor.java:156\u001b[m)
    \u001b[1mat\u001b[m org.apache.maven.lifecycle.internal.MojoExecutor.execute (\u001b[1mMojoExecutor.java:148\u001b[m)
    \u001b[1mat\u001b[m org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (\u001b[1mLifecycleModuleBuilder.java:117\u001b[m)
    \u001b[1mat\u001b[m org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (\u001b[1mLifecycleModuleBuilder.java:81\u001b[m)
    \u001b[1mat\u001b[m org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (\u001b[1mSingleThreadedBuilder.java:56\u001b[m)
    \u001b[1mat\u001b[m org.apache.maven.lifecycle.internal.LifecycleStarter.execute (\u001b[1mLifecycleStarter.java:128\u001b[m)
    \u001b[1mat\u001b[m org.apache.maven.DefaultMaven.doExecute (\u001b[1mDefaultMaven.java:305\u001b[m)
    \u001b[1mat\u001b[m org.apache.maven.DefaultMaven.doExecute (\u001b[1mDefaultMaven.java:192\u001b[m)
    \u001b[1mat\u001b[m org.apache.maven.DefaultMaven.execute (\u001b[1mDefaultMaven.java:105\u001b[m)
    \u001b[1mat\u001b[m org.apache.maven.cli.MavenCli.execute (\u001b[1mMavenCli.java:956\u001b[m)
    \u001b[1mat\u001b[m org.apache.maven.cli.MavenCli.doMain (\u001b[1mMavenCli.java:288\u001b[m)
    \u001b[1mat\u001b[m org.apache.maven.cli.MavenCli.main (\u001b[1mMavenCli.java:192\u001b[m)
    \u001b[1mat\u001b[m jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (\u001b[1mNative Method\u001b[m)
    \u001b[1mat\u001b[m jdk.internal.reflect.NativeMethodAccessorImpl.invoke (\u001b[1mNativeMethodAccessorImpl.java:62\u001b[m)
    \u001b[1mat\u001b[m jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (\u001b[1mDelegatingMethodAccessorImpl.java:43\u001b[m)
    \u001b[1mat\u001b[m java.lang.reflect.Method.invoke (\u001b[1mMethod.java:566\u001b[m)
    \u001b[1mat\u001b[m org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (\u001b[1mLauncher.java:282\u001b[m)
    \u001b[1mat\u001b[m org.codehaus.plexus.classworlds.launcher.Launcher.launch (\u001b[1mLauncher.java:225\u001b[m)
    \u001b[1mat\u001b[m org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (\u001b[1mLauncher.java:406\u001b[m)
    \u001b[1mat\u001b[m org.codehaus.plexus.classworlds.launcher.Launcher.main (\u001b[1mLauncher.java:347\u001b[m)
[\u001b[1;31mERROR\u001b[m] 
[\u001b[1;31mERROR\u001b[m] Re-run Maven using the \u001b[1m-X\u001b[m switch to enable full debug logging.
[\u001b[1;31mERROR\u001b[m] 
[\u001b[1;31mERROR\u001b[m] For more information about the errors and possible solutions, please read the following articles:
[\u001b[1;31mERROR\u001b[m] \u001b[1m[Help 1]\u001b[m http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
container ecbe8af2aeb671f93014b7bd29a33db9d09a27122cdac6f83e7098dcfa856a51 is dead
Thu Jul 30 09:26:48 CEST 2020
```

g4s8 commented:
@rultor release, tag=`0.3.2`

rultor commented:
> @rultor release, tag=`0.3.2`

@g4s8 OK, I will release it now. Please check the progress [here](https://www.rultor.com/t/22766-666188680)

rultor commented:
> @rultor release, tag=`0.3.2`

@g4s8 Oops, I failed. You can see the full log [here](https://www.rultor.com/t/22766-666188680) (spent 23min)

```
    \u001b[1mat\u001b[m org.apache.maven.lifecycle.internal.MojoExecutor.execute (\u001b[1mMojoExecutor.java:210\u001b[m)
    \u001b[1mat\u001b[m org.apache.maven.lifecycle.internal.MojoExecutor.execute (\u001b[1mMojoExecutor.java:156\u001b[m)
    \u001b[1mat\u001b[m org.apache.maven.lifecycle.internal.MojoExecutor.execute (\u001b[1mMojoExecutor.java:148\u001b[m)
    \u001b[1mat\u001b[m org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (\u001b[1mLifecycleModuleBuilder.java:117\u001b[m)
    \u001b[1mat\u001b[m org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (\u001b[1mLifecycleModuleBuilder.java:81\u001b[m)
    \u001b[1mat\u001b[m org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (\u001b[1mSingleThreadedBuilder.java:56\u001b[m)
    \u001b[1mat\u001b[m org.apache.maven.lifecycle.internal.LifecycleStarter.execute (\u001b[1mLifecycleStarter.java:128\u001b[m)
    \u001b[1mat\u001b[m org.apache.maven.DefaultMaven.doExecute (\u001b[1mDefaultMaven.java:305\u001b[m)
    \u001b[1mat\u001b[m org.apache.maven.DefaultMaven.doExecute (\u001b[1mDefaultMaven.java:192\u001b[m)
    \u001b[1mat\u001b[m org.apache.maven.DefaultMaven.execute (\u001b[1mDefaultMaven.java:105\u001b[m)
    \u001b[1mat\u001b[m org.apache.maven.cli.MavenCli.execute (\u001b[1mMavenCli.java:956\u001b[m)
    \u001b[1mat\u001b[m org.apache.maven.cli.MavenCli.doMain (\u001b[1mMavenCli.java:288\u001b[m)
    \u001b[1mat\u001b[m org.apache.maven.cli.MavenCli.main (\u001b[1mMavenCli.java:192\u001b[m)
    \u001b[1mat\u001b[m jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (\u001b[1mNative Method\u001b[m)
    \u001b[1mat\u001b[m jdk.internal.reflect.NativeMethodAccessorImpl.invoke (\u001b[1mNativeMethodAccessorImpl.java:62\u001b[m)
    \u001b[1mat\u001b[m jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (\u001b[1mDelegatingMethodAccessorImpl.java:43\u001b[m)
    \u001b[1mat\u001b[m java.lang.reflect.Method.invoke (\u001b[1mMethod.java:566\u001b[m)
    \u001b[1mat\u001b[m org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (\u001b[1mLauncher.java:282\u001b[m)
    \u001b[1mat\u001b[m org.codehaus.plexus.classworlds.launcher.Launcher.launch (\u001b[1mLauncher.java:225\u001b[m)
    \u001b[1mat\u001b[m org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (\u001b[1mLauncher.java:406\u001b[m)
    \u001b[1mat\u001b[m org.codehaus.plexus.classworlds.launcher.Launcher.main (\u001b[1mLauncher.java:347\u001b[m)
\u001b[1mCaused by\u001b[m: com.sonatype.nexus.staging.client.StagingRuleFailuresException: \u001b[1;31mStaging rules failure!\u001b[m
    \u001b[1mat\u001b[m com.sonatype.nexus.staging.client.internal.StagingWorkflowV3ServiceImpl.monitorRepositoryTransition (\u001b[1mStagingWorkflowV3ServiceImpl.java:205\u001b[m)
    \u001b[1mat\u001b[m com.sonatype.nexus.staging.client.internal.StagingWorkflowV2ServiceImpl.monitorRepositoryTransition (\u001b[1mStagingWorkflowV2ServiceImpl.java:407\u001b[m)
    \u001b[1mat\u001b[m com.sonatype.nexus.staging.client.internal.StagingWorkflowV2ServiceImpl$10.perform (\u001b[1mStagingWorkflowV2ServiceImpl.java:281\u001b[m)
    \u001b[1mat\u001b[m com.sonatype.nexus.staging.client.internal.StagingWorkflowV2ServiceImpl$10.perform (\u001b[1mStagingWorkflowV2ServiceImpl.java:276\u001b[m)
    \u001b[1mat\u001b[m com.sonatype.nexus.staging.client.internal.ExceptionConverter.runAndReturn (\u001b[1mExceptionConverter.java:38\u001b[m)
    \u001b[1mat\u001b[m com.sonatype.nexus.staging.client.internal.StagingWorkflowV2ServiceImpl.releaseStagingRepositories (\u001b[1mStagingWorkflowV2ServiceImpl.java:275\u001b[m)
    \u001b[1mat\u001b[m com.sonatype.nexus.staging.client.internal.StagingWorkflowV3ServiceImpl.releaseStagingRepositories (\u001b[1mStagingWorkflowV3ServiceImpl.java:91\u001b[m)
    \u001b[1mat\u001b[m org.sonatype.nexus.maven.staging.workflow.ReleaseStageRepositoryMojo.doExecute (\u001b[1mReleaseStageRepositoryMojo.java:55\u001b[m)
    \u001b[1mat\u001b[m org.sonatype.nexus.maven.staging.workflow.AbstractStagingActionMojo.execute (\u001b[1mAbstractStagingActionMojo.java:55\u001b[m)
    \u001b[1mat\u001b[m org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (\u001b[1mDefaultBuildPluginManager.java:137\u001b[m)
    \u001b[1mat\u001b[m org.apache.maven.lifecycle.internal.MojoExecutor.execute (\u001b[1mMojoExecutor.java:210\u001b[m)
    \u001b[1mat\u001b[m org.apache.maven.lifecycle.internal.MojoExecutor.execute (\u001b[1mMojoExecutor.java:156\u001b[m)
    \u001b[1mat\u001b[m org.apache.maven.lifecycle.internal.MojoExecutor.execute (\u001b[1mMojoExecutor.java:148\u001b[m)
    \u001b[1mat\u001b[m org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (\u001b[1mLifecycleModuleBuilder.java:117\u001b[m)
    \u001b[1mat\u001b[m org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (\u001b[1mLifecycleModuleBuilder.java:81\u001b[m)
    \u001b[1mat\u001b[m org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (\u001b[1mSingleThreadedBuilder.java:56\u001b[m)
    \u001b[1mat\u001b[m org.apache.maven.lifecycle.internal.LifecycleStarter.execute (\u001b[1mLifecycleStarter.java:128\u001b[m)
    \u001b[1mat\u001b[m org.apache.maven.DefaultMaven.doExecute (\u001b[1mDefaultMaven.java:305\u001b[m)
    \u001b[1mat\u001b[m org.apache.maven.DefaultMaven.doExecute (\u001b[1mDefaultMaven.java:192\u001b[m)
    \u001b[1mat\u001b[m org.apache.maven.DefaultMaven.execute (\u001b[1mDefaultMaven.java:105\u001b[m)
    \u001b[1mat\u001b[m org.apache.maven.cli.MavenCli.execute (\u001b[1mMavenCli.java:956\u001b[m)
    \u001b[1mat\u001b[m org.apache.maven.cli.MavenCli.doMain (\u001b[1mMavenCli.java:288\u001b[m)
    \u001b[1mat\u001b[m org.apache.maven.cli.MavenCli.main (\u001b[1mMavenCli.java:192\u001b[m)
    \u001b[1mat\u001b[m jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (\u001b[1mNative Method\u001b[m)
    \u001b[1mat\u001b[m jdk.internal.reflect.NativeMethodAccessorImpl.invoke (\u001b[1mNativeMethodAccessorImpl.java:62\u001b[m)
    \u001b[1mat\u001b[m jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (\u001b[1mDelegatingMethodAccessorImpl.java:43\u001b[m)
    \u001b[1mat\u001b[m java.lang.reflect.Method.invoke (\u001b[1mMethod.java:566\u001b[m)
    \u001b[1mat\u001b[m org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (\u001b[1mLauncher.java:282\u001b[m)
    \u001b[1mat\u001b[m org.codehaus.plexus.classworlds.launcher.Launcher.launch (\u001b[1mLauncher.java:225\u001b[m)
    \u001b[1mat\u001b[m org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (\u001b[1mLauncher.java:406\u001b[m)
    \u001b[1mat\u001b[m org.codehaus.plexus.classworlds.launcher.Launcher.main (\u001b[1mLauncher.java:347\u001b[m)
[\u001b[1;31mERROR\u001b[m] 
[\u001b[1;31mERROR\u001b[m] Re-run Maven using the \u001b[1m-X\u001b[m switch to enable full debug logging.
[\u001b[1;31mERROR\u001b[m] 
[\u001b[1;31mERROR\u001b[m] For more information about the errors and possible solutions, please read the following articles:
[\u001b[1;31mERROR\u001b[m] \u001b[1m[Help 1]\u001b[m http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
container b482d21ca00647d04df30442c0a7c73ebd8684a9ffebacc31660318709d25288 is dead
Thu Jul 30 09:53:10 CEST 2020
```
